#10539 graph.rst doc edit

Konntroll opened the pull request:
Done with editing of graph.rst for task #10539.

I feel inclined to replace the code snippet in section "Edge and node renderers" with the example linked in the following section "Layout providers" and rewrite both sections to account for this change, merging them into one.

It feels like we are making the user look for what he or she wants to do with the code in two places instead of one. This change might make for a slightly better user experience.

On the other hand, this could overload the section with information, so I'm holding off on such brazen changes. Let me know what you think. Worth it?

Conversely, I could merge the sections and introduce the 2D part of the code as a stand-alone snippet separated from the original one in "Edge and node renderers" by an transition paragraph. This would make the chapter look less intimidating and also address the issue above.

```
### start of layout code
circ = [i*2*math.pi/8 for i in node_indices]
x = [math.cos(i) for i in circ]
y = [math.sin(i) for i in circ]

graph_layout = dict(zip(node_indices, zip(x, y)))
graph.layout_provider = StaticLayoutProvider(graph_layout=graph_layout)

plot.renderers.append(graph)

output_file('graph.html')
show(plot)
```

The code snippet preceding the 2D part in the first of the two sections doesn't map one to one onto the code example mentioned in the second section, though the differences seem to be subtle and/or trivial.

mattpap commented:
Another PR where integration tests failed (this hasn't been happening in a while).

bryevdv commented:
@Konntroll I don't mind merging in to one section since as you point out it's a bit weird to have a first section that doesn't actually end with something usable. That said the first (complete) code block would be quite long. What you do you think about merging in to on section but splitting the full code up in to a couple of sections that have explanatory text around them, instead of one giant block?  Usually we try to have code blocks be "complete" but I think this might be a reasonable exception. cc @tcmetzger 

> Another PR where integration tests failed (this hasn't been happening in a while).

I hadn't noticed myself but will take your word. Did you restart these?

mattpap commented:
> I hadn't noticed myself but will take your word.

I guess it was a temporary hiccup. Three PRs failed and now it's all OK.

tcmetzger commented:
> Usually we try to have code blocks be "complete" but I think this might be a reasonable exception. cc @tcmetzger
I think it absolutely makes sense to break things up a bit at times. You could consider first explaining the code step by step with snippets, and then end the paragraph with the completed code in one big code block. This might seem a bit redundant, but I think that it might actually help to make a complex code example easiert to understand.

Konntroll commented:
@bryevdv & @tcmetzger, alright, I'll give it a go. Might have some questions on how to best explain each of the code blocks, though. We'll see.

Btw, which of the code snippets do you think is preferable, the one from the first section or the one from the example? Like I mentioned, they differ slightly in terms of imports and assignments.

I was thinking of splitting the 2D part off from the main body of code but breaking it up and explaining what each block does as Timo suggests feels even more friendly if indeed slightly redundant.

Konntroll commented:
Split the code snippet in the 'Edge and node renderers' section, added the Cartesian coordinates and output snippet from the example and merged the 'Layout providers' section into the preceding one, (over?)simplifying it in the process.

Not sure how I feel about this setup. Maybe just adding the missing code piece after an introductory paragraph separating it from the declarative part of the program would be a better option after all. If that's the case, I'll revert and rewrite.

Also note that I set example code output to ``none`` because:
* It doesn't match the code in the article, though I'm pretty sure the two do the same thing.
* Felt like there would too much code.

bryevdv commented on sphinx/source/docs/user_guide/graph.rst:
```diff
@@ -44,111 +40,157 @@ Here's a code snippet that:
+* List the nodes and initialize a plot.
+
+.. code-block:: python
+
```
@Konntroll these code snippets currently look like this:

<img width="604" alt="Screen Shot 2020-10-31 at 7 58 43 PM" src="https://user-images.githubusercontent.com/1078448/97794169-a538ef80-1bb3-11eb-9000-a15343889f46.png">

Can you indent the code blocks so that they are under the bullets? I find that much easier to follow. 

Alternatively we could move some of these bullets into code comments and not break up into so many pieces (I had only imagined 2-3 code blocks really)

Konntroll replied:
I agree, it does look fractured and comments make more sense here than bullets.

Revised as originally discussed: two code snippets (with comments) separated by a brief paragraph on the ``LayoutProvider`` model and plot output to close off the section.

Let me know if you'd like any further changes.

bryevdv commented on sphinx/source/docs/user_guide/graph.rst:
```diff
@@ -44,111 +43,139 @@ Here's a code snippet that:
     from bokeh.models import GraphRenderer, Ellipse
     from bokeh.palettes import Spectral8
 
+    ### list the nodes and initialize a plot
```
We don't use a triple hash `###` anywhere else in the docs or examples, I think it should just be `#` (with a single space after) for consistency

bryevdv commented:
Hi @Konntroll I pushed a whitespace linter fix before I noticed the extra hashes in the comments. So you will need to do a pull before making that change.

FYI many editors can be configured to remove trailing whitespace on save (I have VScode configured that way e.g.) Also if you want to run the codebase tests locally before pushing, it is:
```
pytest test/codebase
```
at the top level.

Konntroll commented:
Thanks for pointing out the tests. I believe @mattpap mentioned those to me earlier and I did run some specific ones with mixed results. I also configured PyCharm to trim trailing whitespace, but I do some of my edits on GitHub when at work, so some gets through.

I pulled this branch and fixed the comments, however 5 out of 13 tests fail: eslint with assertion error, the rest with 'file not found'. I'm not sure why. Will try poking around some more before pushing my changes.

bryevdv commented:
@Konntroll  I am also seeing elsint errors locally that are not on CI. @mattpap did something change?

mattpap commented:
@Konntroll, @bryevdv nothing I'm aware of. Let's start with removing `node_modules` and installing and rebuilding from scratch. If the issue persists, I will need to know node/npm versions, platform and exact output.

mattpap commented:
I also updated bokehjs' dependencies in PR #10634, but that didn't reveal anything.

bryevdv commented:
@Konntroll I'd say go ahead and push changes if the the only local issue is the eslint one

bryevdv commented:
Thanks @Konntroll !
